### ISSUE TITLE
Implementing Pattern Recognition for OCLL

### ISSUE THREAD

**Issue by adiwastu** [tdecker91/puzzle-gen#3]
![image](https://user-images.githubusercontent.com/19344345/232947189-0d998ab2-78d9-4382-ac74-e4c717673e36.png)

For OLL cases where the edges are all oriented, there are 7 possible cases, and above is one of them (commonly named "U" OCLL). The full 7 OCLL names would be U, T, L, H, Pi, S, AS. There can be 4 possible angles for this but they're relatively the same, so accounting for a U/U'/U2 turns from that position, an OCLL can have 4 different angles, and working from there, the 7 initial OCLL states can be expanded to 7 * 4 = 28 shapes.

Is it possible for PuzzleGen to output some kind of text/ store some text in a variable that shows what OCLL cases you generated? Strictly for the oriented edges case, so we can just check the color of the corners that faces up. Like for example,

1. Check how many corner that has the upper layer corner that faces up.
2. If there is none, it has to be H or Pi. check the orientation of the rest of the corners to decide.
3. If there is 1, it's has to be S or AS. Again, check the orientation of the rest of the corners to decide.
4. If there is 2, check if the corners are adjacent of each other. If it is opposite, it can only be L. If they're adjacent, check the orientation of the rest of the corners to decide.

Will something like that be possible to implement? Thank you for your time for this!

Best,
Adi

**Comment by tdecker91** [comment]
Thanks for the feedback. 

It _is_ certainly possible and I'm considering what is the best way to expose that. I think I'd like to know the use case a little bit more before knowing how to handle it.

For some context, one of my goals with this library is to provide the puzzle rendering capability and be pretty agnostic to use case so the library can be utilized in different applications.

My first thought is that the application asking for an OLL render already knows the case in mind. So perhaps just providing a way to add a text label to the image itself would be sufficient?

Or perhaps the application is looking at a random scramble and trying to detect the case, I think this is more appropriate to include the logic outside of the puzzle-gen library, but expose the internal state of the puzzle so it can be used. I'm assuming this is more what you're after?

The good news is that the puzzle state is already exposed somewhat.

The modules are arranged kind of like this

- rendering
  - PNG
  - SVG
  - Canvas
- visualizer
  - simulator
  - renderer

with `PNG` `SVG` `Canvas` just convenience methods to build a visualizer and output the image to the HTML document. Which is the core purpose of the library. A `visualizer` keeps the state of the puzzle in the `simulator` and then uses w/e renderer to output depending on the mode selected (png, svg, canvas)

the `SVG` and `Canvas` methods return the visualizer object and the puzzle state can be inspected to do what you're looking for.

**Comment by adiwastu** [comment]
Hey, thank you! I would love to explain my use case by quoting your comments and perhaps explain how mine might be a bit too niche:

> My first thought is that the application asking for an OLL render already knows the case in mind.

So what I have now is around 25k of algs that I'm sure what it does (it affects OLL) but not quite sure what OLL it gave. The algs are generated by combining 2 algs, and that is why the case is not known to the application. I tried bruteforcing the algs, but could only handle filtering images for like a couple hundred combinations before getting tired (lol). That brings me to my next small code, basically I used puzzlegen.min.js to do the work of displaying the algs!

```<html>
  <head>
    <script type="text/javascript" src="puzzleGen.min.js"></script>
  </head>
  <body>
    <div>
    // some div ...
    </div>
    <script type="text/javascript">
      function generatePuzzles(algorithmsId, containerId) {
        let algorithms = document.getElementById(algorithmsId).value.split("\n");
        let puzzleContainer = document.getElementById(containerId);
        puzzleContainer.innerHTML = "";

        for (let i = 0; i < algorithms.length; i++) {
          let options = {
            "width": 100,
            "height": 100,
            "strokeWidth": 0.03,
            puzzle:
            {
              alg: algorithms[i]
            }
          };
          let puzzleDiv = document.createElement("div");
          puzzleDiv.id = "cube" + (i + 1);
          puzzleContainer.appendChild(puzzleDiv);
          puzzleGen.PNG("#" + puzzleDiv.id, puzzleGen.Type.CUBE, options);

          // Add an indicator text to show the current index
          let indicatorText = document.createElement("p");
          indicatorText.innerHTML = "Generating puzzle " + (i + 1) + " of " + algorithms.length;
          puzzleDiv.appendChild(indicatorText);
        }
      }
    </script>
  </body>
</html>
```

As you can see, a super simple one. If I could somehow pull the state of the cube (like what color UFR is, or basically every sticker) I could very easily make logic to differentiate the case. And as you said here:

> the `SVG` and `Canvas` methods return the visualizer object and the puzzle state can be inspected to do what you're looking for.

could you please guide me on how to pull those state from the objects? Thanks a bunch!

**Comment by adiwastu** [comment]
Hi Tyson,

Looks like I've found the way to expose the stickers that could work—by messing around on the console and from your lead (visualizer). This code shows the map and I am able to use the get() method to extract the stickers that matter, with `LL_state` being an object that is returned by the `Canvas` method.

```LL_state.simulator.stickers;```

Closing this issue. I might document what I did when my project is done, and for whoever reading this and find this pokes your interest, please expect some proof of concept codes on my page in the future. Thank you!

Best,
Adi

**Comment by tdecker91** [comment]
Yup exactly. 
Sorry, I was planning to write out a full example / description on the weekend when I had some time. But you get the idea. 

To be honest it might not be very straightforward but you should be able to accomplish what you need. The simulator stuff wasn't really built with that in mind, but if it works that's great!

**Comment by adiwastu** [comment]
Your lead worked wonders nevertheless! And I did get that done just several hours ago. Since I only cared about the corners of the last layers, that is 4 * 3 stickers, I pulled those info and evaluated whether they got the "U color" value. And then it's just the matter of joining them and comparing them to some literals (the 7 * 4 OCLL patterns/angles) and that's it! [Here's a pic:](https://cdn.discordapp.com/attachments/714831722122051645/1101416957997228042/image.png)

Cheers!
